src/test/java/reactor/pool/CommonPoolTest.java

```diff
@@ -531,15 +525,90 @@ void smokeTestAsyncLifo(Function<PoolBuilder<PoolableTest, ?>, AbstractPool<Pool
+			throws InterruptedException {
+		AtomicInteger allocationCount = new AtomicInteger();
+
+		PoolBuilder<PoolableTest, ?> builder = PoolBuilder
+				.from(Mono.fromCallable(() -> {
+					int id = allocationCount.incrementAndGet();
+					return new PoolableTest(id);
+				}))
+				.sizeBetween(4, 8)
+				.releaseHandler(p -> Mono.fromRunnable(p::clean));
+
+		final Pool<PoolableTest> pool = configAdjuster.apply(builder);
+		assertThat(pool).isInstanceOf(InstrumentedPool.class);
+
+		final PoolMetrics metrics = ((InstrumentedPool) pool).metrics();
+		assertThat(metrics.allocatedSize()).as("constructor allocated").isEqualTo(4);
+		assertThat(metrics.idleSize()).as("warmup idle").isEqualTo(4);
+
+		final PooledRef<PoolableTest> firstAcquire = pool.acquire()
+		                                                  .block();
+
+		assertThat(metrics.allocatedSize()).as("first acquire allocated").isEqualTo(4);
+		assertThat(metrics.idleSize()).as("first acquire idle").isEqualTo(3);
+
+		//give time for an unexpected warmup to take place so that the last assertion is valid
+		Thread.sleep(100);
+		assertThat(allocationCount).as("total allocations").hasValue(4);
+	}
+
+	@ParameterizedTest
+	@MethodSource("allPools")
+	void destroyBelowMinSizeThreshold(Function<PoolBuilder<PoolableTest, ?>, Pool<PoolableTest>> configAdjuster)
+			throws InterruptedException {
+		AtomicInteger allocationCount = new AtomicInteger();
+
+		PoolBuilder<PoolableTest, ?> builder = PoolBuilder
+				.from(Mono.fromCallable(() -> {
+					int id = allocationCount.incrementAndGet();
+					return new PoolableTest(id);
+				}))
+				.sizeBetween(3, 4)
+				.releaseHandler(p -> Mono.fromRunnable(p::clean));
+
+		final Pool<PoolableTest> pool = configAdjuster.apply(builder);
+		assertThat(pool).isInstanceOf(InstrumentedPool.class);
+		final PoolMetrics metrics = ((InstrumentedPool) pool).metrics();
+
+		final PooledRef<PoolableTest> ref1 = pool.acquire().block();
+		final PooledRef<PoolableTest> ref2 = pool.acquire().block();
+		final PooledRef<PoolableTest> ref3 = pool.acquire().block();
+		final PooledRef<PoolableTest> ref4 = pool.acquire().block();
+
+		assertThat(metrics.allocatedSize()).as("initial allocated").isEqualTo(4);
+		assertThat(metrics.idleSize()).as("initial idle").isEqualTo(0);
+
+		ref1.invalidate().block();
+		ref2.invalidate().block();
+		ref3.invalidate().block();
+
+		assertThat(metrics.allocatedSize()).as("3 releases: allocated").isEqualTo(1);
+		assertThat(metrics.idleSize()).as("3 releases: idle").isEqualTo(0);
+
+		PooledRef<PoolableTest> ref5 = pool.acquire()
+		                                   .block();
+
+		assertThat(metrics.allocatedSize()).as("second acquire allocated").isEqualTo(3); //1 borrowed, 1 acquiring, 1 extra matches min of 3
+		assertThat(metrics.idleSize()).as("second acquire idle").isEqualTo(1); //only 1 extra created this time
+
+		//give time for an unexpected warmup to take place so that the last assertion is valid
+		Thread.sleep(100);
```

Comment:
timing sensitive, how about Awaitility?

Reply:
the sleep here is a buffer for an unexpected warmup to asynchronously take place, but the asserted count should ALREADY be at the desired value at the start of the sleep. we want to ensure it stays there even after a small delay. `Awaitility`'s `atLeast` doesn't seem to fit the bill here. The javadoc is a bit unclear, but the wiki says:
> In case the condition is fulfilled before the duration specified by atLeast an exception is thrown indicating that the condition shouldn't be completed earlier than the specified duration.

I didn't find an obvious method to check that, so `sleep` sound like the best most readable option.